libbeat/processors/uuid/generator/elasticsearch/generator.go

```diff
@@ -0,0 +1,111 @@
+// ID. The algorithm used to generate the ID is the same as used by Elasticsearch.
+// See https://github.com/elastic/elasticsearch/blob/a666fb2266/server/src/main/java/org/elasticsearch/common/TimeBasedUUIDGenerator.java
+func GetBase64UUID() string {
+	mu.Lock()
+	defer mu.Unlock()
+
+	sequenceNumber++
+
+	// We only use bottom 3 bytes for the sequence number.
+	s := sequenceNumber & 0xffffff
+
+	timestamp := getTimestamp()
+	if s == 0 {
+		// Always force the clock to increment whenever sequence number is 0, in case we have a long time-slip backwards.
+		timestamp.Add(1 * time.Millisecond)
+	}
+	lastTimestamp = timestamp
+
+	t := timestamp.UnixNano() / 1000 // timestamp in ms-since-epoch
+
+	uuidBytes := make([]byte, 15)
+
+	//// We have auto-generated ids, which are usually used for append-only workloads.
+	//// So we try to optimize the order of bytes for indexing speed (by having quite
+	//// unique bytes close to the beginning of the ids so that sorting is fast) and
+	//// compression (by making sure we share common prefixes between enough ids).
+
+	// We use the sequence number rather than the timestamp because the distribution of
+	// the timestamp depends too much on the indexing rate, so it is less reliable.
+	uuidBytes[0] = byte(s)       // copy lowest-order byte from sequence number
+	uuidBytes[1] = byte(s >> 16) // copy 3rd lowest-order byte from sequence number
+
+	// Now we start focusing on compression and put bytes that should not change too often.
+	uuidBytes[2] = byte(t >> 16) // 3rd lowest-order byte from timestamp; changes every ~65 secs
+	uuidBytes[3] = byte(t >> 24) // 4th lowest-order byte from timestamp; changes every ~4.5h
+	uuidBytes[4] = byte(t >> 32) // 5th lowest-order byte from timestamp; changes every ~50 days
+	uuidBytes[5] = byte(t >> 40) // 6th lowest-order byte from timestamp; changes every 35 years
+
+	// Copy mac address bytes (6 bytes)
+	copy(uuidBytes[6:6+addrLen], mac)
+
+	// Finally we put the remaining bytes, which will likely not be compressed at all.
+	uuidBytes[12] = byte(t >> 8) // 2nd lowest-order byte from timestamp
+	uuidBytes[13] = byte(s >> 8) // 2nd lowest-order byte from sequence number
+	uuidBytes[14] = byte(t)
+
+	// See also: more detailed explanation of byte choices at
+	// https://github.com/elastic/elasticsearch/blob/a666fb22664284d8e2114841ebb58ea4e1924691/server/src/main/java/org/elasticsearch/common/TimeBasedUUIDGenerator.java#L80-L95
+
+	return base64.RawURLEncoding.EncodeToString(uuidBytes)
+}
+
+func getTimestamp() *time.Time {
+	// Don't let timestamp go backwards, at least "on our watch" (while this process is running).  We are still vulnerable if we are
+	// shut down, clock goes backwards, and we restart... for this we randomize the sequenceNumber on init to decrease chance of
+	// collision.
+	now := time.Now()
+
+	if lastTimestamp == nil {
+		return &now
+	}
+
+	if lastTimestamp.After(now) {
```

Comment:
:) I wonder how reliable this check is. `time.Time` stores two timestamps. The wall clock and the monotonic clock (if available). It uses the monotonic clock when comparing times. Its very likely that this condition is false always.

From docs:

```
In addition to the required “wall clock” reading, a Time may contain an optional reading of the current process's monotonic clock, to provide additional precision for comparison or subtraction. See the “Monotonic Clocks” section in the package documentation for details.
```

`UnixNano` will be computed based on the wall clock time, whereas `After` or `Before` will compare based on the monotonic clock if a monotonic clock was available. I think you are better of using the UnixNano for lastTimestamp right away.

Unfortunately `UnixNano` can be subject to switch in daylight savings for example. Now 3 bytes for the sequence number doesn't look that nice anymore. Makes me wonder if we should add/update a 'delta' if the difference is > 30min. This way we could give the impression of a more monotonic clock if the machines timezone is suddenly changed.

Reply:
Ah, thanks for the info. I also read up on https://godoc.org/time#hdr-Monotonic_Clocks, and I agree that we want to compare wall clock time here, particularly since we are using `UnixNano` here: https://github.com/elastic/beats/pull/14524/files#diff-99e67f960cab16cacf647698d860514bR62. So given that the monotonic clock is used for comparisons but the wall clock is used for telling the time, we could end up in a scenario where this check doesn't pass (because the montonic clock hasn't gone backward) however we end up using a time in the past anyway (for example, when the wall clock is moved backwards due to a DST change).

Rather than try to add a delta, why not just use the wall clock value for comparison here? So something like this:

```go
	now := time.Now()

	if lastTimestamp == nil {
		return now
	}

        if lastTimestamp.UnixNanos() > now.UnixNanos() {
            return lastTimestamp
        }

        return now
```

Reply:
1. lets move the getter into the other function. The lastTimestamp value is used here, but set in another place. If we deal with globals, then their usage should be in as few functions/places as possible
2. we can change the type of lastTimestamp to uint64. The we can reduce the number of conversations. The first check becomes `lastTimestamp == 0` :)
3. the idea with the delta is to reduce the duration IDs with the same prefix are generated for. For example:

```
const timezoneOffset = 30 * time.Minute 
now := time.Now().UnixNanos()
if lastTimestamp != 0 {
  if lastTimestamp > now {
    if (lastTimestamp - now) > timezoneOffset {
      // looks like someone did adjust the timezone. Let's compute a delta we will
      // add to future timestamps.
      delta = ((lastTimestamp - now) / (timezoneOffset+1)) * timezoneOffset // should we round up to next hour?
      if lastTimestamp > now {
        return lastTimestamp
      }

      lastTimestamp = now
      return now + delta
    }
    return lastTimestamp + delta
  } else if (now - lastTimestamp) >= delta {
    // reset delta if it looks like we either didn't create an ID for a long time
    // or if the timezone has been set back (beats running >6month?)
    delta = 0
  }
}

lastTimestamp = now
return now + delta
```

If someone sets the clock back by ~1h we will have a delta of `60min`. In the future we add the delta to all timestamps, making it appear like the clock was not changed this much.

Well, functionality like this must be well tested :/

Reply:
@urso After reading through the above code, I'm wondering if it needs to be specific to the DST change use case. 

I think there's a more general problem here, that is if the system time (`now`) falls behind the last timestamp computed (`lastTimestamp`) for _whatever_ reason (could be DST change, could be something else). I've implemented a solution that I think solves the more general problem using a delta like you showed above, but without requiring the `timezoneOffset`. Maybe I'm misunderstanding the problem though, so would you mind giving my solution a look? It's in https://github.com/elastic/beats/pull/14524/commits/e4f9fe9b8e75e3b723a5d1074aa4b9a460ddde85. Thanks!